### DISCUSSION TITLE
Run multiple instances of immich docker

### DISCUSSION THREAD

**Discussion by iceisraw** [immich-app/immich#375] [Q&A]
Hi. I tried the immich using the docker compose. Thank you very much for this app. We liked it so much that others in my household want to use the same app but with isolation from one another. Such that each of them have their docker app running. I want to run each instance of this app through docker on one Ubuntu host. Can each instance of the app run such that they use own separate backup folder and their own ip/port. Is it easily possible ? I tried simply running same thing twice with env file modified for each to account for port and folder path but it did not work. I am not sure if what I am asking even makes sense or considered appropriate from docker perspective.

**Comment by alextran1502** [comment]
What environment are you hosting, an RPi, or a dedicated server? If you have a dedicated server, you can spin up multiple VMs with different IP addresses so that each user will have access to their instance.

Suppose you want to spin up multiple instances on the same machine. In that case, it could hamper the machine's performance if multiple users choose to upload simultaneously, triggering image and video conversion (heavy CPU tasks). So to do that, follow the steps below:

1. Create a separate folder for each instance with a distinct name
2. Create a `docker-compose.yml` and put the `docker-compose` file content into it ([link](https://github.com/alextran1502/immich/blob/27ea3fc881/docker/docker-compose.yml))
3. In the docker-compose file, `immich-proxy`, use a different port for each instance.

For example, for Peter

```yaml
  immich-proxy:
    container_name: immich_proxy
    image: altran1502/immich-proxy:release
    ports:
      - 2284:80
    logging:
      driver: none
    depends_on:
      - immich-server
    restart: always
```

for Alex

```yaml
  immich-proxy:
    container_name: immich_proxy
    image: altran1502/immich-proxy:release
    ports:
      - 2285:80
    logging:
      driver: none
    depends_on:
      - immich-server
    restart: always
```
5. Populate the `.env` file in each folder with a different `UPLOAD_LOCATION`, which is dedicated to each person and the port you specified above for `VITE_SERVER_ENDPOINT`
6. From each folder, run `docker-compose up -d`

Anyway, I suggest using one instance with multiple accounts.

**Comment by iceisraw** [comment]
Hi Alex thanks for detailed help. Now I tried to start a second one but it
errored while trying to create the immich_postgres and immich_redis .
Cannot create due to a conflict same name and port already in use.( I think
this is due to first instance also using same port for Postgres and redis)
. Is there a way I can add custom port myself for these? Again thank you so
much for your help and time.

**Reply by alextran1502** [reply]
What is your folder structure? You probably have to create a custom network for each instance, so they don't run into each other.

https://docs.docker.com/compose/networking/

**Comment by iceisraw** [comment]
Simply changing the port in env file is not effective for these Postgres
and redis containers. It does work in case if immich-proxy but for that we
add port in both env and yml file.

**Reply by sohail579** [reply]
did you ever get this working?

**Comment by VegetableElk** [comment]
Here is an example of how to run an e.g. test version of immich side by side with a production version.

Note the use of `links:` in docker-compose.yml.

TLDR; rename the containers so they don't clash, then add network dns aliases to them with the `links:` setting so that the in-code dns lookups resolve to the differently named containers within the internal network


```
version: "3.8"

#
# WARNING: Make sure to use the docker-compose.yml of the current release:
#
# https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml
#
# The compose file on main may not be compatible with the latest release.
#

name: immich2

services:
  immich-server2:
    container_name: immich_server2
    links:
      - redis2:immich_redis
      - database2:immich_postgres
      - immich-microservices2:immich-microservices
      - immich-machine-learning2:immich-machine-learning
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${EXTERNAL_PATH}:/usr/src/app/external:ro
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    ports:
      - 2284:3001
    depends_on:
      - redis2
      - database2
    restart: always

  immich-microservices2:
    container_name: immich_microservices2
    links:
      - redis2:immich_redis
      - database2:immich_postgres
      - immich-machine-learning2:immich-machine-learning
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/hardware-transcoding
      file: hwaccel.transcoding.yml
      service: quicksync # set to one of [nvenc, quicksync, rkmpp, vaapi, vaapi-wsl] for accelerated transcoding
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${EXTERNAL_PATH}:/usr/src/app/external:ro
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    depends_on:
      - redis2
      - database2
    restart: always

  immich-machine-learning2:
    container_name: immich_machine_learning2
    # For hardware acceleration, add one of -[armnn, cuda, openvino] to the image tag.
    # Example tag: ${IMMICH_VERSION:-release}-cuda
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}-openvino
    extends: # uncomment this section for hardware acceleration - see https://immich.app/docs/features/ml-hardware-acceleration
       file: hwaccel.ml.yml
       service: openvino # set to one of [armnn, cuda, openvino, openvino-wsl] for accelerated inference - use the `-wsl` version for WSL2 where applicable
    volumes:
      - model-cache2:/cache
    env_file:
      - .env
    restart: always

  redis2:
    container_name: immich_redis2
    image: redis:6.2-alpine@sha256:51d6c56749a4243096327e3fb964a48ed92254357108449cb6e23999c37773c5
    restart: always

  database2:
    container_name: immich_postgres2
    image: tensorchord/pgvecto-rs:pg14-v0.2.0@sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      - pgdata2:/var/lib/postgresql/data
    restart: always

volumes:
  pgdata2:
  model-cache2:

```

**Reply by ghan1t** [reply]
If anyone is coming across this discussion, I had to add the following two lines to each links section to make it work with the current version of immich:
```
      - redis2:redis
      - database2:database
```
Server and microservices try to reach "redis" and "database" as dns names on the docker network and the link alias to immich_redis/immich_postgres was not enough.

**Reply by ochen1** [reply]
As of v1.112.1 you'll need:
```yml
    links:
      - redis2:redis
      - database2:database
      - immich2-machine-learning:immich-machine-learning
```

**Comment by codedcosmos** [comment]
As of writing you don't need to update links anymore, it doesn't exist in the docker compose. I did however need to set the environment variable REDIS_HOSTNAME and DB_HOSTNAME

Since I encountered the same issue as this person https://github.com/immich-app/immich/discussions/9533

Environment var docs https://immich.app/docs/install/environment-variables/
